File: app/javascript/controllers/test_controller.js
import { Controller } from "@hotwired/stimulus"

export default class extends Controller {
  static targets = ["submit"]

    connect() {
    console.log("Stimulus: TestController connected")
    }

  loading() {
    console.log("Stimulus: loading triggered")
    this.submitTarget.disabled = true
    this.submitTarget.textContent = "Loading..."
  }

  success(event) {
    const [data, _status, _xhr] = event.detail
    this.flashMessage(data.message, "notice")
    this.reset()
  }

  error(event) {
    const [data, _status, _xhr] = event.detail
    this.flashMessage(data.message || "Something went wrong", "alert")
    this.reset()
  }

  reset() {
    this.submitTarget.disabled = false
    this.submitTarget.textContent = "Click Me"
  }

  flashMessage(message, type) {
    let flash = document.querySelector(".flash")
    if (!flash) {
      flash = document.createElement("div")
      flash.classList.add("flash", type)
      document.body.prepend(flash)
    }
    flash.textContent = message
    flash.className = `flash ${type}`
  }
}

File: app/controllers/test_controller.rb
class TestController < ApplicationController
  def index
    # render view
  end

  def create
    sleep 2

    respond_to do |format|
      format.html do
        flash[:notice] = "✅ Created at #{Time.now.strftime("%H:%M:%S")}"
        redirect_to test_index_path
      end

      format.json do
        render json: { message: "✅ Created at #{Time.now.strftime("%H:%M:%S")}" }
      end
    end
  rescue => e
    Rails.logger.error "Create failed: #{e.message}"

    respond_to do |format|
      format.html do
        flash[:alert] = "Something went wrong. #{e.message}"
        redirect_to test_index_path
      end

      format.json do
        render json: { message: "Something went wrong." }, status: :unprocessable_entity
      end
    end
  end
end
